src/main/java/org/openrewrite/staticanalysis/OnlyCatchDeclaredExceptions.java

```diff
@@ -0,0 +1,136 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://docs.moderne.io/licensing/moderne-source-available-license
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.staticanalysis;
+
+import org.jspecify.annotations.Nullable;
+import org.openrewrite.*;
+import org.openrewrite.internal.ListUtils;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.JavaTemplate;
+import org.openrewrite.java.ShortenFullyQualifiedTypeReferences;
+import org.openrewrite.java.search.UsesType;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaType;
+import org.openrewrite.java.tree.JavaType.FullyQualified;
+import org.openrewrite.java.tree.TypeUtils;
+
+import java.util.*;
+
+import static java.util.stream.Collectors.joining;
+
+public class OnlyCatchDeclaredExceptions extends Recipe {
+
+    private static final String JAVA_LANG_EXCEPTION = "java.lang.Exception";
+
+    @Override
+    public String getDisplayName() {
+        return "Replace `catch(Exception)` with specific declared exceptions thrown in the try block";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Replaces `catch(Exception e)` blocks with a multi-catch block " +
+                "(`catch (SpecificException1 | SpecificException2 e)`) containing only the exceptions declared " +
+                "thrown by method or constructor invocations within the `try` block that are not already caught " +
+                "by more specific `catch` clauses.";
+    }
+
+    @Override
+    public Set<String> getTags() {
+        return new HashSet<>(Arrays.asList("CWE-396", "RSPEC-S2221"));
+    }
+
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        JavaIsoVisitor<ExecutionContext> visitor = new JavaIsoVisitor<ExecutionContext>() {
+            @Override
+            public J.Try visitTry(J.Try aTry, ExecutionContext ctx) {
+                J.Try t = super.visitTry(aTry, ctx);
+                return t.withCatches(ListUtils.map(t.getCatches(), c -> {
+                    if (isGenericCatch(c)) {
+                        // Find declared thrown exceptions that are not already specifically caught
+                        Set<JavaType> declaredThrown = getDeclaredThrownExceptions(t);
+                        declaredThrown.removeAll(getCaughtExceptions(t));
+                        if (!declaredThrown.isEmpty()) {
+                            return multiCatchWithDeclaredExceptions(c, declaredThrown);
+                        }
+                    }
+                    return c;
+                }));
```

Comment:
```suggestion
                // Find declared thrown exceptions that are not already specifically caught
                Set<JavaType> declaredThrown = getDeclaredThrownExceptions(t);
                declaredThrown.removeAll(getCaughtExceptions(t));
                return t.withCatches(ListUtils.map(t.getCatches(), c -> {
                    if (isGenericCatch(c) && !declaredThrown.isEmpty()) {
                            return multiCatchWithDeclaredExceptions(c, declaredThrown);
                    }
                    return c;
                }));
```

To avoid iterating over the try block multiple times?